nni/contrib/compression/base/compressor.py

```diff
@@ -0,0 +1,300 @@
+
+
+class Compressor:
+    def __init__(self, model: torch.nn.Module, config_list: List[Dict], mode: Literal['pruning', 'quantization', 'distillation'],
+                 evaluator: Evaluator | None = None, existed_wrappers: Dict[str, ModuleWrapper] | None = None):
+        """
+        Compressor base class.
+
+        Parameters
+        ----------
+        model
+            Model to be compressed.
+        config_list
+            Config list. TODO: please refer.
+        mode
+            One of ['pruning', 'quantization', 'distillation'] compress mode.
+        evaluator
+            TODO: please refer.
+        existed_wrappers
+            Use by class method ``from_compressor`` to inherit another compressor's wrapper.
+        """
+        assert mode in ['pruning', 'quantization', 'distillation']
+        self.bound_model = model
+        self.config_list = trans_legacy_config_list(deepcopy(config_list))
+        self._validate_config()
+        self.evaluator = evaluator
+        if self.evaluator is not None:
+            assert isinstance(evaluator, Evaluator)
+            if not evaluator._initialization_complete:
+                evaluator._init_optimizer_helpers(self.bound_model)
+
+        self._is_wrapped = False
+        self._module_wrappers, self._target_spaces = register_wrappers(self.bound_model, self.config_list, mode, existed_wrappers)
+        self.wrap_model()
+
+    @classmethod
+    def from_compressor(cls, compressor: Compressor, new_config_list: List[Dict], *args, evaluator: Evaluator | None = None, **kwargs):
+        """
+        Inherited from the compressor exsited wrappers and evaluator to initialize a new compressor.
+
+        Parameters
+        ----------
+        compressor
+            The bound model, wrappers and evaluator in this compressor will be inherited to create a new compressor.
+        new_config_list
+            The config_list used to config the new compressor.
+        evaluator
+            Key-word only parameter. If the inherited compressor doesn't have an evaluator, then this evaluator will be used to
+            initialize the new compressor. If the inherited compressor already has an evaluator, this parameter will be ignored.
+        args
+            Positional arguments. Will be directly passed to the ``__init__`` function.
+        kwargs
+            Keyword arguments. Will be directly passed to the ``__init__`` function.
+        """
+        if compressor._is_wrapped:
+            compressor.unwrap_model()
+        model = compressor.bound_model
+        existed_wrappers = compressor._module_wrappers
+        if compressor.evaluator is not None and evaluator is not None:
+            _logger.warning('compessor already has evaluator, the new evaluator passed to this function will be ignored.')
+        evaluator = compressor.evaluator if compressor.evaluator else evaluator
+
+        # note that here don't have `mode` because subclass should know what its mode is.
```

Comment:
a little confused, where is the mode parameter defined? Can you give an example of a subclass?

Reply:
yes, it's confused, I will del `mode` from init interface, instead, add it as a class attr.